R/interpolate.R

```diff
@@ -0,0 +1,359 @@
+#' @param interpolate_vals \[`numeric(1)`\]\cr
+#'   The values of `interpolate_col` for which you would like to solve for
+#'   an interpolated value of `value_col`.
+#'
+#' @details This function uses `stats::approx` to solve a linear interpolation.
+#'   Values outside the bounds of known data will be returned as NA. Use
+#'   [demUtils::extrapolate()] to get values outside of the bounds of the data.
+#'   Consider log transforming your data prior to interpolation if
+#'   appropriate.
+#'
+#' @return \[`data.table()`\] `dt` with added rows for interpolated values.
+#'
+#' @examples
+#' dt <- data.table::data.table(
+#'   group = c(rep("a", 5), rep("b", 5)),
+#'   x = c(1, 3, 4, 5, 10, 1, 2, 6, 8, 10),
+#'   y = c(10, 30, 40, 50, 100, 10, 20, 60, 80, 100)
+#' )
+#' dt <- interpolate(dt, id_cols = c("group", "x"), interpolate_col = "x",
+#'                   value_col = "y", interpolate_vals = c(1:10))
+#'
+#' @export
+interpolate <- function(dt, id_cols, interpolate_col, value_col,
+                        interpolate_vals) {
+
+  dt <- copy(dt)
+
+  # validate ----------------------------------------------------------------
+
+  # check `id_cols`
+  assertthat::assert_that(is.character(id_cols))
+
+  # check `dt`
+  assertable::assert_colnames(
+    dt, c(id_cols, value_col), only_colnames = F, quiet = T
+  )
+
+  # check `interpolate_col`
+  assertthat::assert_that(
+    assertthat::is.string(interpolate_col),
+    interpolate_col %in% id_cols,
+    is.numeric(dt[, get(interpolate_col)]),
+    msg = paste0("`interpolate_col` must be an entry of `id_cols` and ",
+                 "represent a numeric column of `dt`.")
+  )
+
+  # check `value_col`
+  assertthat::assert_that(
+    assertthat::is.string(value_col),
+    value_col %in% names(dt),
+    is.numeric(dt[, get(value_col)]),
+    msg = "`value_col` must represent a numeric column of `dt`."
+  )
+
+  # check `interpolate_vals`
+  assertthat::assert_that(is.numeric(interpolate_vals))
+  lb <- min(dt[, get(interpolate_col)])
+  ub <- max(dt[, get(interpolate_col)])
+  assertthat::assert_that(
+    between(min(interpolate_vals), lb, ub) |
+      between(max(interpolate_vals), lb, ub),
+    msg = paste0("There is no overlap between the range of your data and ",
+                 "the range of `interpolate_vals`.")
+  )
+
+  # expand ------------------------------------------------------------------
+
+  # fill in NAs
+  expand_cols <- list(temp = interpolate_vals)
+  names(expand_cols) <- interpolate_col
+  dt <- expand(dt, expand_cols = expand_cols, id_cols = id_cols)
+
+  # interpolate -------------------------------------------------------------
+
+  dt <- dt[order(get(interpolate_col))]
+  dt <- rbindlist(
+    lapply(
+      split(dt, by = setdiff(id_cols, interpolate_col)),
+      FUN = function(d) {
+        d$temp <- stats::approx(
+          x = d[, get(interpolate_col)],
+          y = d[, get(value_col)],
+          xout = d[, get(interpolate_col)]
+        )$y
```

Comment:
maybe make `method` an argument that we can pass on to `stats::approx` and just inherit documentation

Reply:
Or use `...` to pass on any arguments to `approx`